File: 1254-number-of-closed-islands/1254-number-of-closed-islands.kt
data class Coordinate(val x: Int, val y: Int)

class Solution {
    val dx = listOf(0, 1, 0, -1)
    val dy = listOf(1, 0, -1, 0)

    fun closedIsland(grid: Array<IntArray>): Int {
        val visited = Array(grid.size) { IntArray(grid[0].size) { Integer.MAX_VALUE } }
        val n = grid.size
        val m = grid[0].size
        var result = 0

        for (i in 0 until n) {
            for (j in 0 until m) {
                if (!isWall(i, j, grid) && visited[i][j] == Integer.MAX_VALUE) {
                    if (bfs(i, j, grid, visited, n, m) != -1) result++
                }
            }
        }

        return result
    }

    private fun bfs(x: Int, y: Int, grid: Array<IntArray>, visited: Array<IntArray>, n: Int, m: Int): Int {
        val start = Coordinate(x, y)
        visited[x][y] = 1
        val q: Queue<Coordinate> = LinkedList<Coordinate>()
        q.add(start)

        val boundaryCells = mutableListOf<Coordinate>()

        while(q.isNotEmpty()) {
            val current = q.poll()

            for (dir in 0 until 4) {
                val nx = current.x + dx[dir]
                val ny = current.y + dy[dir]

                if (isNotClosedIsland(nx, ny, n, m)) {
                    boundaryCells.add(current)
                    continue
                }
                if (isWall(nx, ny, grid)) continue
                val nextDistance = visited[current.x][current.y] + 1
                if (visited[nx][ny] > nextDistance) {
                    visited[nx][ny] = nextDistance
                    q.add(Coordinate(nx, ny))
                }
            }
        }

        for (cell in boundaryCells) {
            if (!isWall(cell.x, cell.y, grid)) {
                return -1
            }
        }

        return 0
    }

    private fun isNotClosedIsland(x: Int, y: Int, n: Int, m: Int) = x < 0 || x >= n || y < 0 || y >= m

    private fun isWall(x: Int, y: Int, grid: Array<IntArray>) = grid[x][y] == 1
}

File: 1254-number-of-closed-islands/1254-number-of-closed-islands.cpp
class Solution {
private:
    int dx[4] = {-1, 0, 1, 0};
    int dy[4] = {0, 1, 0, -1};

    int n;
    int m;
    bool visited[100][100] = { 0, };

    bool isInBoundary(int x, int y) {
        if (0 <= x && x < n && 0 <= y && y < m) return true;
        return false;
    }

    bool isClosedIsland(int x, int y, vector<vector<int>> &grid) {
        queue<pair<int, int>> q;
        q.push({x, y});
        visited[x][y] = true;

        bool isClosedIsland = true;

        while(!q.empty()) {
            int cx = q.front().first;
            int cy = q.front().second;
            q.pop();

            for (int dir = 0; dir < 4; dir++) {
                int nx = cx + dx[dir];
                int ny = cy + dy[dir];
                if (!isInBoundary(nx, ny)) {
                    isClosedIsland = false;
                    continue;
                }
                if (grid[nx][ny] == 1 || visited[nx][ny]) continue;

                visited[nx][ny] = true;
                q.push({nx, ny});
            }
        }
        return isClosedIsland;
    }
public:
    int closedIsland(vector<vector<int>>& grid) {
        n = grid.size();
        m = grid[0].size();

        int result = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 1 || visited[i][j]) continue;
                if (isClosedIsland(i, j, grid)) result++;
            }
        }

        return result;
    }
};

